File: frontend/src/app/services/offer-letter.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { 
  OfferLetter, 
  OfferLetterRequest, 
  OfferLetterResponse,
  EligibleCandidate 
} from '../models/offer-letter.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class OfferLetterService {
  private apiUrl = `${environment.apiUrl}/offer-letters`;

  constructor(private http: HttpClient) { }

  private getHeaders(): HttpHeaders {
    return new HttpHeaders({
      'Content-Type': 'application/json'
    });
  }

  /**
   * Generate new offer letter
   */
  generateOfferLetter(request: OfferLetterRequest): Observable<OfferLetterResponse> {
    return this.http.post<OfferLetterResponse>(
      `${this.apiUrl}/generate`, 
      request,
      { headers: this.getHeaders() }
    );
  }

  /**
   * Get all offer letters
   */
  getAllOfferLetters(): Observable<OfferLetterResponse[]> {
    return this.http.get<OfferLetterResponse[]>(this.apiUrl);
  }

  /**
   * Get offer letter by ID
   */
  getOfferLetterById(id: string): Observable<OfferLetterResponse> {
    return this.http.get<OfferLetterResponse>(`${this.apiUrl}/${id}`);
  }

  /**
   * Get offer letter by candidate ID
   */
  getOfferLetterByCandidateId(candidateId: string): Observable<OfferLetterResponse> {
    return this.http.get<OfferLetterResponse>(`${this.apiUrl}/candidate/${candidateId}`);
  }

  /**
   * Get offer letters by status
   */
  getOfferLettersByStatus(status: string): Observable<OfferLetterResponse[]> {
    return this.http.get<OfferLetterResponse[]>(`${this.apiUrl}/status/${status}`);
  }

  /**
   * Send offer letter to candidate
   */
  sendOfferLetter(id: string): Observable<OfferLetterResponse> {
    return this.http.post<OfferLetterResponse>(
      `${this.apiUrl}/${id}/send`,
      {},
      { headers: this.getHeaders() }
    );
  }

  /**
   * Accept offer letter
   */
  acceptOfferLetter(id: string, acceptanceMethod?: string): Observable<OfferLetterResponse> {
    return this.http.post<OfferLetterResponse>(
      `${this.apiUrl}/${id}/accept`,
      { acceptanceMethod: acceptanceMethod || 'PORTAL' },
      { headers: this.getHeaders() }
    );
  }

  /**
   * Reject offer letter
   */
  rejectOfferLetter(id: string, rejectionReason: string): Observable<OfferLetterResponse> {
    return this.http.post<OfferLetterResponse>(
      `${this.apiUrl}/${id}/reject`,
      { rejectionReason },
      { headers: this.getHeaders() }
    );
  }

  /**
   * Download offer letter
   */
  downloadOfferLetter(id: string): Observable<OfferLetterResponse> {
    return this.http.get<OfferLetterResponse>(`${this.apiUrl}/${id}/download`);
  }

  /**
   * Get candidates eligible for offer letter
   */
  getEligibleCandidates(): Observable<EligibleCandidate[]> {
    return this.http.get<EligibleCandidate[]>(`${this.apiUrl}/eligible-candidates`);
  }

  /**
   * Calculate salary breakdown based on CTC
   */
  calculateSalaryBreakdown(ctc: number): {
    basic: number,
    hra: number,
    specialAllowance: number,
    performanceBonus: number
  } {
    const basic = Math.round(ctc * 0.40); // 40% of CTC
    const hra = Math.round(ctc * 0.20); // 20% of CTC
    const specialAllowance = Math.round(ctc * 0.30); // 30% of CTC
    const performanceBonus = Math.round(ctc * 0.10); // 10% of CTC
    
    return {
      basic,
      hra,
      specialAllowance,
      performanceBonus
    };
  }

  /**
   * Format currency
   */
  formatCurrency(amount: number, currency: string = 'INR'): string {
    if (currency === 'INR') {
      return `₹${amount.toLocaleString('en-IN')}`;
    }
    return `${currency} ${amount.toLocaleString()}`;
  }

  /**
   * Get status badge class
   */
  getStatusBadgeClass(status: string): string {
    const statusMap: { [key: string]: string } = {
      'GENERATED': 'badge-info',
      'SENT': 'badge-primary',
      'ACCEPTED': 'badge-success',
      'REJECTED': 'badge-danger',
      'EXPIRED': 'badge-warning',
      'WITHDRAWN': 'badge-secondary'
    };
    return statusMap[status] || 'badge-secondary';
  }
}

File: backend/src/main/java/com/smarthire/service/OfferLetterService.java
package com.smarthire.service;

import com.smarthire.dto.OfferLetterRequest;
import com.smarthire.dto.OfferLetterResponse;
import com.smarthire.model.CandidateApplication;
import com.smarthire.model.Interview;
import com.smarthire.model.Job;
import com.smarthire.model.OfferLetter;
import com.smarthire.model.ShortlistedCandidate;
import com.smarthire.repository.CandidateApplicationRepository;
import com.smarthire.repository.InterviewRepository;
import com.smarthire.repository.JobRepository;
import com.smarthire.repository.OfferLetterRepository;
import com.smarthire.repository.ShortlistedCandidateRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class OfferLetterService {
    
    private final OfferLetterRepository offerLetterRepository;
    private final InterviewRepository interviewRepository;
    private final ShortlistedCandidateRepository shortlistedCandidateRepository;
    private final CandidateApplicationRepository applicationRepository;
    private final JobRepository jobRepository;
    
    /**
     * Generate offer letter for candidate who cleared all interviews
     */
    public OfferLetterResponse generateOfferLetter(OfferLetterRequest request) {
        try {
            // Validate that candidate cleared HR round
            Interview hrInterview = interviewRepository.findById(request.getInterviewId())
                    .orElseThrow(() -> new RuntimeException("Interview not found"));
            
            if (!"HR_ROUND".equals(hrInterview.getInterviewRound())) {
                throw new RuntimeException("Candidate has not completed HR round");
            }
            
            if (!"SELECTED".equals(hrInterview.getDecision())) {
                throw new RuntimeException("Candidate not selected in HR round");
            }
            
            // Check if offer letter already exists
            if (offerLetterRepository.findByApplicationId(request.getApplicationId()).isPresent()) {
                throw new RuntimeException("Offer letter already generated for this candidate");
            }
            
            // Get candidate details
            ShortlistedCandidate candidate = shortlistedCandidateRepository
                    .findById(request.getCandidateId())
                    .orElseThrow(() -> new RuntimeException("Candidate not found"));
            
            // Get application details
            CandidateApplication application = applicationRepository
                    .findById(request.getApplicationId())
                    .orElseThrow(() -> new RuntimeException("Application not found"));
            
            // Get job details
            Job job = jobRepository.findById(request.getJobId())
                    .orElseThrow(() -> new RuntimeException("Job not found"));
            
            // Create offer letter
            OfferLetter offerLetter = new OfferLetter();
            offerLetter.setCandidateId(request.getCandidateId());
            offerLetter.setApplicationId(request.getApplicationId());
            offerLetter.setJobId(request.getJobId());
            offerLetter.setInterviewId(request.getInterviewId());
            
            // Candidate information
            offerLetter.setCandidateName(candidate.getCandidateName());
            offerLetter.setCandidateEmail(candidate.getCandidateEmail());
            offerLetter.setCandidatePhone(application.getPhone());
            
            // Job information
            offerLetter.setJobTitle(job.getTitle());
            offerLetter.setDepartment(null); // Department not available in Job model
            offerLetter.setCompany(job.getCompany());
            
            // Offer details
            offerLetter.setOfferLetterNumber("OL-" + System.currentTimeMillis());
            offerLetter.setOfferDate(LocalDate.now());
            offerLetter.setJoiningDate(request.getJoiningDate());
            offerLetter.setEmploymentType(request.getEmploymentType() != null ? request.getEmploymentType() : "FULL_TIME");
            offerLetter.setWorkLocation(request.getWorkLocation() != null ? request.getWorkLocation() : job.getLocation());
            offerLetter.setOfficeLocation(request.getOfficeLocation() != null ? request.getOfficeLocation() : job.getLocation());
            
            // Compensation
            offerLetter.setAnnualCtc(request.getAnnualCtc());
            offerLetter.setBasicSalary(request.getBasicSalary());
            offerLetter.setHra(request.getHra());
            offerLetter.setSpecialAllowance(request.getSpecialAllowance());
            offerLetter.setPerformanceBonus(request.getPerformanceBonus());
            offerLetter.setOtherAllowances(request.getOtherAllowances());
            offerLetter.setCurrency("INR");
            
            // Benefits
            offerLetter.setBenefits(request.getBenefits());
            offerLetter.setPaidLeaves(request.getPaidLeaves() != null ? request.getPaidLeaves() : 24);
            offerLetter.setAdditionalBenefits(request.getAdditionalBenefits());
            
            // Terms
            offerLetter.setProbationPeriod(request.getProbationPeriod() != null ? request.getProbationPeriod() : 3);
            offerLetter.setNoticePeriod(request.getNoticePeriod() != null ? request.getNoticePeriod() : 30);
            offerLetter.setReportingManager(request.getReportingManager());
            offerLetter.setReportingManagerDesignation(request.getReportingManagerDesignation());
            
            // Status
            offerLetter.setStatus("GENERATED");
            offerLetter.setGeneratedAt(LocalDateTime.now());
            LocalDate expiryDateValue = request.getExpiryDate() != null ? request.getExpiryDate() : LocalDate.now().plusDays(15);
            offerLetter.setExpiryDate(expiryDateValue);
            
            // Additional
            offerLetter.setAdditionalNotes(request.getAdditionalNotes());
            offerLetter.setGeneratedBy(request.getGeneratedBy());
            offerLetter.setApprovedBy(request.getApprovedBy());
            offerLetter.setIsAccepted(false);
            offerLetter.setIsDownloaded(false);
            offerLetter.setDownloadCount(0);
            offerLetter.setUpdatedAt(LocalDateTime.now());
            
            // Save offer letter
            OfferLetter saved = offerLetterRepository.save(offerLetter);
            
            // Update candidate status in shortlisted collection
            candidate.setOfferLetterGenerated(true);
            candidate.setOfferLetterId(saved.getId());
            candidate.setStatus(ShortlistedCandidate.ShortlistStatus.OFFER_LETTER_SENT);
            shortlistedCandidateRepository.save(candidate);
            
            return mapToResponse(saved, "Offer letter generated successfully");
            
        } catch (Exception e) {
            throw new RuntimeException("Failed to generate offer letter: " + e.getMessage());
        }
    }
    
    /**
     * Get all offer letters
     */
    public List<OfferLetterResponse> getAllOfferLetters() {
        return offerLetterRepository.findAll().stream()
                .map(ol -> mapToResponse(ol, null))
                .collect(Collectors.toList());
    }
    
    /**
     * Get offer letter by ID
     */
    public OfferLetterResponse getOfferLetterById(String id) {
        OfferLetter offerLetter = offerLetterRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offer letter not found"));
        return mapToResponse(offerLetter, null);
    }
    
    /**
     * Get offer letter by candidate ID
     */
    public OfferLetterResponse getOfferLetterByCandidateId(String candidateId) {
        OfferLetter offerLetter = offerLetterRepository.findByCandidateId(candidateId)
                .orElseThrow(() -> new RuntimeException("No offer letter found for this candidate"));
        return mapToResponse(offerLetter, null);
    }
    
    /**
     * Get offer letters by status
     */
    public List<OfferLetterResponse> getOfferLettersByStatus(String status) {
        return offerLetterRepository.findByStatus(status).stream()
                .map(ol -> mapToResponse(ol, null))
                .collect(Collectors.toList());
    }
    
    /**
     * Send offer letter to candidate
     */
    public OfferLetterResponse sendOfferLetter(String id) {
        OfferLetter offerLetter = offerLetterRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offer letter not found"));
        
        offerLetter.setStatus("SENT");
        offerLetter.setSentAt(LocalDateTime.now());
        offerLetter.setUpdatedAt(LocalDateTime.now());
        
        OfferLetter updated = offerLetterRepository.save(offerLetter);
        
        // TODO: Send email to candidate with offer letter
        
        return mapToResponse(updated, "Offer letter sent successfully");
    }
    
    /**
     * Accept offer letter
     */
    public OfferLetterResponse acceptOfferLetter(String id, String acceptanceMethod) {
        OfferLetter offerLetter = offerLetterRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offer letter not found"));
        
        if (!"SENT".equals(offerLetter.getStatus())) {
            throw new RuntimeException("Offer letter must be in SENT status to accept");
        }
        
        offerLetter.setStatus("ACCEPTED");
        offerLetter.setIsAccepted(true);
        offerLetter.setRespondedAt(LocalDateTime.now());
        offerLetter.setAcceptanceMethod(acceptanceMethod != null ? acceptanceMethod : "PORTAL");
        offerLetter.setUpdatedAt(LocalDateTime.now());
        
        OfferLetter updated = offerLetterRepository.save(offerLetter);
        
        // Update candidate status
        ShortlistedCandidate candidate = shortlistedCandidateRepository
                .findById(offerLetter.getCandidateId())
                .orElseThrow(() -> new RuntimeException("Candidate not found"));
        candidate.setStatus(ShortlistedCandidate.ShortlistStatus.OFFER_ACCEPTED);
        shortlistedCandidateRepository.save(candidate);
        
        return mapToResponse(updated, "Offer letter accepted successfully");
    }
    
    /**
     * Reject offer letter
     */
    public OfferLetterResponse rejectOfferLetter(String id, String rejectionReason) {
        OfferLetter offerLetter = offerLetterRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offer letter not found"));
        
        offerLetter.setStatus("REJECTED");
        offerLetter.setIsAccepted(false);
        offerLetter.setRespondedAt(LocalDateTime.now());
        offerLetter.setRejectionReason(rejectionReason);
        offerLetter.setUpdatedAt(LocalDateTime.now());
        
        OfferLetter updated = offerLetterRepository.save(offerLetter);
        
        // Update candidate status
        ShortlistedCandidate candidate = shortlistedCandidateRepository
                .findById(offerLetter.getCandidateId())
                .orElseThrow(() -> new RuntimeException("Candidate not found"));
        candidate.setStatus(ShortlistedCandidate.ShortlistStatus.OFFER_REJECTED);
        shortlistedCandidateRepository.save(candidate);
        
        return mapToResponse(updated, "Offer letter rejected");
    }
    
    /**
     * Download offer letter (increment download count)
     */
    public OfferLetterResponse downloadOfferLetter(String id) {
        OfferLetter offerLetter = offerLetterRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Offer letter not found"));
        
        offerLetter.setIsDownloaded(true);
        offerLetter.setDownloadCount(offerLetter.getDownloadCount() + 1);
        offerLetter.setUpdatedAt(LocalDateTime.now());
        
        OfferLetter updated = offerLetterRepository.save(offerLetter);
        
        return mapToResponse(updated, "Offer letter downloaded");
    }
    
    /**
     * Get candidates eligible for offer letter (cleared HR round)
     */
    public List<ShortlistedCandidate> getCandidatesEligibleForOfferLetter() {
        // Find all HR round interviews with SELECTED decision
        List<Interview> hrInterviews = interviewRepository
                .findByInterviewRoundAndDecision("HR_ROUND", "SELECTED");
        
        // Get shortlisted candidates who don't have offer letter yet
        return hrInterviews.stream()
                .map(interview -> shortlistedCandidateRepository
                        .findById(interview.getShortlistedCandidateId())
                        .orElse(null))
                .filter(candidate -> candidate != null && 
                        (candidate.getOfferLetterGenerated() == null || !candidate.getOfferLetterGenerated()))
                .collect(Collectors.toList());
    }
    
    /**
     * Helper method to map OfferLetter to Response
     */
    private OfferLetterResponse mapToResponse(OfferLetter offerLetter, String message) {
        OfferLetterResponse response = new OfferLetterResponse();
        response.setId(offerLetter.getId());
        response.setOfferLetterNumber(offerLetter.getOfferLetterNumber());
        response.setCandidateName(offerLetter.getCandidateName());
        response.setCandidateEmail(offerLetter.getCandidateEmail());
        response.setJobTitle(offerLetter.getJobTitle());
        response.setCompany(offerLetter.getCompany());
        response.setOfferDate(offerLetter.getOfferDate());
        response.setJoiningDate(offerLetter.getJoiningDate());
        response.setEmploymentType(offerLetter.getEmploymentType());
        response.setWorkLocation(offerLetter.getWorkLocation());
        response.setAnnualCtc(offerLetter.getAnnualCtc());
        response.setCurrency(offerLetter.getCurrency());
        response.setStatus(offerLetter.getStatus());
        response.setGeneratedAt(offerLetter.getGeneratedAt());
        response.setSentAt(offerLetter.getSentAt());
        response.setIsAccepted(offerLetter.getIsAccepted());
        response.setOfferLetterPdfUrl(offerLetter.getOfferLetterPdfUrl());
        response.setIsDownloaded(offerLetter.getIsDownloaded());
        response.setMessage(message);
        return response;
    }
}
